syntax = "proto3";

package grpc;

message Empty {}

// I called it Note because message Message sounds complicated
message Note { // must be remove
    string name = 1;
    string message = 2;
}

message StartedBool{
    bool started = 1;
}

message PrivateInfo{
    bytes ip = 1;
    string user = 2;
}

message Ping {
    bytes ip = 1;
}

message Pong {
    string message = 1;
}

message GameStatus {
    string timeStamp = 1;
    string playerPositions = 2; // stringified array of position
    string lastPlayer = 3; // sender of messge    
}

message InitialList {
    int32 length = 1;
    string name_hp = 2; // ["utente1","utente2"][hp1, hp2]
    string player_type = 3; //["1","0"]
    string ip = 4; //["enc_ip","enc_ip"]
}

message Health{
    bytes ip = 1;
    int32 hp = 2; 
    string user = 3;
}

message Block{
    bytes ip = 1;
    int32 block = 2; 
    string user = 3;
}

message Action{
    bytes ip_sender = 1;
    bytes ip_reciever = 2;
    int32 amount = 3;
    int32 action_type = 4; //0 = attack, 1 = heal, 2 = block etc etc
}

service ChatServer {
    // This bi-directional stream makes it possible to send and receive Notes between 2 persons
    rpc ChatStream (Empty) returns (stream Note);
    rpc SendNote (Note) returns (Empty);
    rpc SendPrivateInfo (PrivateInfo) returns (Empty);
    rpc SendPing(Ping) returns (Pong);
    rpc StartGame(PrivateInfo) returns(InitialList);
    rpc SendHealth(Health) returns(Empty);
    rpc SendBlock(Block) returns(Empty);
    rpc SendAction(Action) returns(Empty);
    rpc HealthStream(Empty) returns(stream Health);
    rpc BlockStream(Empty) returns(stream Block);
    rpc ActionStream(Empty) returns(stream Action);
    rpc GetInitialList(Empty) returns(InitialList);
    rpc FinishGame(Empty) returns(Note);
    rpc EndTurn(PrivateInfo) returns(Empty);
    rpc TurnStream (Empty) returns(stream PrivateInfo);
    rpc ReturnStarted (Empty) returns(StartedBool);
}
// python3 -m grpc_tools.protoc -I=. --python_out=. --grpc_python_out=. chat.proto 
// python -m grpc_tools.protoc -I. --python_out=. --grpc_python_out=. chat.proto      //Simone is a very special boy